pkg/archivetool/archivetool.go

```diff
@@ -49,6 +49,7 @@ func PluginArchiveToTempDir(uri string) (string, func(), error) {
 	// validate is a dir
 	fileInfo, err := os.Stat(archivePath)
 	if err != nil {
+		archiveCleanup()
```

Comment:
Sorry, I think I created some confusion here in my first review, and also by taking a closer look I realised that simply using `defer` won't work as I had suggested in the first review.

`archiveCleanup` needs to be called after it has been initialized, in any case when `PluginArchiveToTempDir` returns an error. If there's no error (err = nil), the caller will have to take care of calling it.

Above the highlighted line, there are two other cases where the function can return an error (lines 42 and 45), and another one below here (line 57). You should either add `archiveCleanup()` before these `return`s as well, or clean up with defer:

```go
defer func() {
	if err != nil && archiveCleanup != nil {
		archiveCleanup()
	}
}()
```

(this `defer` should go right after the first err check, so on line 38)

I would use defer, so it's less repetitive.

If you use `defer`, line 45 should also become:

```go
err = errors.New("no files in archive")
return "", err
```

And similarly, line 57 should become:

```go
err = errors.New("Archive root is not a directory. Plugins archives must contain a single directory at the root")
return "", nil, err
```

Otherwise the error is not set to `err`, and the deferred function will not see it. I glossed over this during my first review, my bad!

Reply:
Got it! thanks